core/shared/src/main/scala/scalaz/zio/Retry.scala

```diff
@@ -0,0 +1,428 @@
+  final def initialized(f: IO[E, self.State] => IO[E, self.State]): Retry[E, S] =
+    new Retry[E, S] {
+      type State = self.State
+      val initial                              = f(self.initial)
+      def proj(state: State): S                = self.proj(state)
+      def update(e: E, s: State): IO[E, State] = self.update(e, s)
+    }
+}
+
+object Retry {
+
+  /**
+   * Constructs a new retry strategy from an initial state and an update function.
+   */
+  final def apply[E, S](initial0: IO[E, S], update0: (E, S) => IO[E, S]): Retry[E, S] = new Retry[E, S] {
+    type State = S
+    val initial                              = initial0
+    def proj(state: State): S                = state
+    def update(e: E, s: State): IO[E, State] = update0(e, s)
+  }
+
+  /**
+   * A retry strategy that always fails.
+   */
+  final def never[E]: Retry[E, Unit] =
+    Retry[E, Unit](IO.unit, (e, _) => IO.fail(e))
+
+  /**
+   * A retry strategy that always succeeds.
+   */
+  final def always[E]: Retry[E, Unit] =
+    Retry[E, Unit](IO.unit, (e, _) => IO.unit)
+
+  /**
+   * A retry strategy that always succeeds with the specified constant state.
+   */
+  final def point[E, S](s: => S): Retry[E, S] =
+    Retry[E, S](IO.point(s), (_, s) => IO.now(s))
+
+  /**
+   * A retry strategy that always succeeds, collecting all errors into a list.
+   */
+  final def errors[E]: Retry[E, List[E]] =
+    Retry[E, List[E]](IO.now(Nil), (e, l) => IO.now(e :: l))
+
+  /**
+   * A retry strategy that always retries and counts the number of retries.
+   */
+  final def counted[E]: Retry[E, Int] =
+    Retry[E, Int](IO.now(0), (_, i) => IO.now(i + 1))
+
+  /**
+   * A retry strategy that always retries and computes the time since the
+   * beginning of the process.
+   */
+  final def elapsed[E]: Retry[E, Duration] = {
+    val nanoTime = IO.sync(System.nanoTime())
+
+    Retry[E, (Long, Long)](nanoTime.seq(IO.now(0L)), (_, t) => nanoTime.map(t2 => (t._1, t2 - t._1)))
+      .map(t => Duration(t._2, TimeUnit.NANOSECONDS))
+  }
+
+  /**
+   * A retry strategy that will keep retrying until the specified number of
+   * retries is reached.
+   */
+  final def retries[E](max: Int): Retry[E, Int] = counted.untilState(_ >= max)
+
+  /**
+   * A retry strategy that will keep retrying until the specified duration has
+   * elapsed.
+   */
+  final def duration[E](duration: Duration): Retry[E, Duration] = {
+    val nanos = duration.toNanos
+
+    elapsed.untilState(_.toNanos >= nanos)
+  }
+
+  /**
+   * A retry strategy that will always succeed, waiting the specified fixed
+   * duration between attempts.
+   */
+  final def fixed[E](duration: Duration): Retry[E, Int] =
+    counted.updated((_, _, io) => io.delay(duration))
+
+  final def backoff[E](base: Duration,
+                       sleep: (Int, Duration) => IO[E, Duration],
+                       cap: Option[Duration] = None): Retry[E, (Int, Duration)] = {
+    val up: ((Int, Duration)) => IO[E, (Int, Duration)] = {
```

Comment:
I would not use `cap` here, because it can be easily added in an orthogonal way using `||`, for example.

This leaves `sleep`. I think we should keep this but push it to `backoff0`, so that `backoff` can use the ordinary (non-jitter) sleep.

We can clean that up later and remove the sleep when we generalize all this to work with arbitrary `F[_, _]`, because we can run it on `Jitter(io)`, for example.

Reply:
Ah, I see, this is the more powerful version of `exponentialBackoff`. So I'd call this one `exponential0`.

Reply:
I thought about that, re: cap. And it doesn't make any sense as an orthogonal concept since it only applies to growing-duration strategies which happen to be only backoff onces.

Reply:
Nonetheless, my point is that you can get this with (`foo || fixed(20.milliseconds)`), so rather than bake the cap in, we should allow the users to add `|| cap` if they want such a cap (it's already an `Option` because not all users will want the cap).